staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go

```diff
@@ -0,0 +1,131 @@
+	"time"
+
+	"github.com/golang/glog"
+
+	"google.golang.org/grpc"
+
+	"golang.org/x/net/context"
+
+	kmsapi "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1"
+)
+
+const (
+	// Now only supportied unix domain socket.
+	unixProtocol = "unix"
+
+	// Current version for the protocal interface definition.
+	kmsapiVersion = "v1beta1"
+)
+
+// The gRPC implementation for envelope.Service.
+type gRPCService struct {
+	// gRPC client instance
+	kmsClient  kmsapi.KMSServiceClient
+	connection *grpc.ClientConn
+}
+
+// NewGRPCService returns an envelope.Service which use gRPC to communicate the remote KMS provider.
+func NewGRPCService(endpoint string) (Service, error) {
+	glog.V(4).Infof("Configure KMS provider with endpoint: %s", endpoint)
+
+	protocol, addr, err := parseEndpoint(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	dialer := func(addr string, timeout time.Duration) (net.Conn, error) {
+		return net.DialTimeout(protocol, addr, timeout)
+	}
+
+	connection, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialer))
+	if err != nil {
+		return nil, fmt.Errorf("connect remote KMS provider %q failed, error: %v", addr, err)
+	}
+
+	kmsClient := kmsapi.NewKMSServiceClient(connection)
+
+	err = checkAPIVersion(kmsClient)
+	if err != nil {
+		connection.Close()
+		return nil, fmt.Errorf("failed check version for %q, error: %v", addr, err)
+	}
+
+	return &gRPCService{kmsClient: kmsClient, connection: connection}, nil
+}
+
+// Parse the endpoint to extract schema, host or path.
+func parseEndpoint(endpoint string) (string, string, error) {
+	if len(endpoint) == 0 {
+		return "", "", fmt.Errorf("remote KMS provider can't use empty string as endpoint")
+	}
+
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", "", fmt.Errorf("invalid endpoint %q for remote KMS provider, error: %v", endpoint, err)
+	}
+
+	if u.Scheme != unixProtocol {
+		return "", "", fmt.Errorf("unsupported scheme %q for remote KMS provider", u.Scheme)
+	}
+	return unixProtocol, u.Path, nil
+}
+
+// Check the KMS provider API version.
+// Only matching kubeRuntimeAPIVersion is supported now.
+func checkAPIVersion(kmsClient kmsapi.KMSServiceClient) error {
+	request := &kmsapi.VersionRequest{Version: kmsapiVersion}
+	response, err := kmsClient.Version(context.Background(), request)
+	if err != nil {
+		return fmt.Errorf("failed get version from remote KMS provider: %v", err)
+	}
+	if response.Version != kmsapiVersion {
+		return fmt.Errorf("KMS provider api version %s is not supported, only %s is supported now",
+			response.Version, kmsapiVersion)
+	}
+
+	glog.V(4).Infof("KMS provider %s initialized, version: %s", response.RuntimeName, response.RuntimeVersion)
+	return nil
+}
+
+// Decrypt a given data string to obtain the original byte data.
+func (g *gRPCService) Decrypt(cipher []byte) ([]byte, error) {
+	request := &kmsapi.DecryptRequest{Cipher: cipher, Version: kmsapiVersion}
+	response, err := g.kmsClient.Decrypt(context.Background(), request)
```

Comment:
I would expect timeouts a timeout of timed cancel on the contexts for these calls.

Reply:
Okay, I will add it.